### PR TITLE
flow: remove discovery_target_decode function from the standard library

### LINKED ISSUE #3607: Flow: replace discovery_target_decode with components
discovery_target_decode was introduced as a way to act as a substitute for the functionality of Prometheus' HTTP and file discovery with `local.file` and `remote.http`. 

However, this approach was flawed: 

* Prometheus file discovery supports YAML files, which cannot be done with local.file/discovery_target_decode. 
* I think it was a mistake to introduce component-specific functions. 

I would like to replace `discovery_target_decode` with two components:

* `discovery.file_targets` for Prometheus' file SD. 
  * `discovery.file` is already in-use for file globbing, unless we renamed that component.  
* `discovery.http` for Prometheus' HTTP SD. 

And going forward, I'd like to change the guidelines around the stdlib to _not_ include any component-specific functions, and keep the stdlib generic. 

I'm looking for feedback on this proposal.

### LINKED ISSUE #4305: discovery_target_decode is considered a component rather than standard lib function
### What's wrong?

The following error is raised when using teh std lib function `discovery_target_decode`:

`Error: /etc/agent-config/agent.river:27:14: component "discovery_target_decode" does not exist `

It seems that discovery_target_decode is considered a component rather a standard lib function. I don't have similar issues with the rest of the functions from the std lib.
I'm running version v0.34.3 of the agent in docker.

### Steps to reproduce

The bug can be reproduced by running the example code found in the documentation: https://grafana.com/docs/agent/latest/flow/reference/stdlib/discovery_target_decode/

```
local.file "example" {
  filename = env("TARGETS_FILE")
}

prometheus.scrape "default" {
  targets = discovery_target_decode(local.file.example.content)
}
```

### System information

Docker

### Software version

v0.34.3

### Configuration

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by rfratto**
Depends on #4404. 

Closes #3607
Closes #4305

### CODE REVIEW

File: pkg/flow/internal/stdlib/stdlib.go [hunk 1]
```diff
@@ -1,46 +0,0 @@
...
-	"github.com/prometheus/prometheus/discovery/targetgroup"
-)
-
-// Identifiers holds a list of stdlib identifiers by name. All interface{}
-// values are River-compatible values.
-//
-// Function identifiers are Go functions with exactly one non-error return
-// value, with an optionally supported error return value as the second return
-// value.
-var Identifiers = map[string]interface{}{
```
Comment (captncraig) [inline pkg/flow/internal/stdlib/stdlib.go:18]: Should we keep this file around for future usages. I know https://github.com/grafana/agent/pull/4388/files uses it for one thing, unless this is the wrong place to add those. I still think its possible there are flow-specific stdlib functions possible.

Comment (rfratto) [reply pkg/flow/internal/stdlib/stdlib.go:18]: #4388 is extending pkg/river/internal/stdlib, but this is removing pkg/flow/internal/stdlib. We consider it an anti-pattern now to have Flow-specific stdlib functions, which is why this file is being removed in favor of only having the River one.

**Review by captncraig — APPROVED** [review]

**Review by mattdurham — APPROVED** [review]
LGTM

**Review by clayton-cornell — APPROVED** [review]
LGTM

**Pull request merged**